[device/celestica/x86_64-cel_silverstone-r0] Fix QSFP-DD support

ds952811 opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

1. This is built on top of PR91, which carries forward the changes in PR91 with some modifications
i.e. This is a replacement of PR91, and you don't need PR91 with this PR merged.
2. switchboard.c: FPGA-I2C: Enable high-speed mode (i.e. BIT6)
3. switchboard.c: FPGA-I2C: Fix the turnaround issues in jiffies comparison
4. switchboard.c: FPGA-I2C: Yield the CPU resources during the busy-wait for I2C ready bit
5. sfputil.py: Revise get_transceiver_change_event() to poll the presence flags and generate change events accordingly, such that SFP insertion events will surely be generated at every boot attempt of 'pmon' instead of physical transceiver insertion/removal.
6. sfputil.py: Add QSFPDD initialization sequence for 'InnoLight T-DP4CNT-N00' to correctly enable 4x100G in case that HwSKU=Silverstone-128x100
7. sfputil.py: Automatically reset the QSFPDD transceiver if its state != ModuleReady for 3+ seconds.
8. sfputil.py: Add support of displaying QSFP-DD application advertisements, and the current application mode.
9. Silverstone-128x100/port_config.ini: Fix the interface naming conflicts (It looks like typos)
10.sfputil.py: Monitoring the QSFP-DD initialization state, and reinitiate it if necessary

**- How I did it**
See **- What I did**

**- How to verify it**
1. Bring up the DUT with the default config, and ensure it's in 'Silverstone-128x100'
2. Attach the QSFP-DD transceivers for the link up/down tests

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

Commit:
Add sysfs for XCVR interrupt event status

Commit:
implement get_transceiver_change_event in sfutil of silverstone

Commit:
add cmis init script

Commit:
update perssion for cmis-init script

Commit:
clean cmis init code

Commit:
clean sfputil code

ds952811 committed:
SilverStone: fix QSFPDD support

Signed-off-by: Dante Su <dante.su@broadcom.com>

ds952811 committed:
skip cmis_init if module absent, add error message upon retry timeout

Signed-off-by: Dante Su <dante.su@broadcom.com>

ds952811 committed:
Silverstone-128x100: fix the naming defect in port_config.ini

Signed-off-by: Dante Su <dante.su@broadcom.com>

ds952811 committed:
sfputil: automatically perform QSFP-DD software reset if module state != ModuleReady for 3+ seconds

Signed-off-by: Dante Su <dante.su@broadcom.com>

ds952811 committed:
sfputil: always perform a software reset upon QSFPDD module insertion

ds952811 commented:
Just a heads up, the following xcvrd crash observed after multiple attempts of QSFPDD reinsertions, and hence we'll also need PR5 of sonic-platform-common for this.

i.e.
https://github.com/zhenggen-xu/sonic-platform-common/pull/5

Nov 18 00:57:55.488171 sonic INFO pmon#supervisord: xcvrd Traceback (most recent call last):
Nov 18 00:57:55.488171 sonic INFO pmon#supervisord: xcvrd File "/usr/bin/xcvrd", line 683, in
Nov 18 00:57:55.488171 sonic INFO pmon#supervisord: xcvrd sys.exit(main())
Nov 18 00:57:55.488171 sonic INFO pmon#supervisord: xcvrd File "/usr/bin/xcvrd", line 606, in main
Nov 18 00:57:55.488188 sonic INFO pmon#supervisord: xcvrd rc = post_port_sfp_info_to_db(logical_port, int_tbl)
Nov 18 00:57:55.488188 sonic INFO pmon#supervisord: xcvrd File "/usr/bin/xcvrd", line 265, in post_port_sfp_info_to_db
Nov 18 00:57:55.488206 sonic INFO pmon#supervisord: xcvrd port_info_dict = platform_sfputil.get_transceiver_info_dict(physical_port)
Nov 18 00:57:55.488206 sonic INFO pmon#supervisord: xcvrd File "/usr/local/lib/python2.7/dist-packages/sonic_platform_base/sonic_sfp/sfputilbase.py", line 726, in get_transceiver_info_dict
Nov 18 00:57:55.501232 sonic INFO pmon#supervisord: xcvrd sfp_vendor_pn_raw = self._read_eeprom_specific_bytes(sysfsfile_eeprom, (offset + OSFP_VENDOR_PN_OFFSET), XCVR_VENDOR_PN_WIDTH)
Nov 18 00:57:55.501232 sonic INFO pmon#supervisord: xcvrd File "/usr/local/lib/python2.7/dist-packages/**sonic_platform_base/sonic_sfp/sfputilbase.py**", line 314, in _read_eeprom_specific_bytes
Nov 18 00:57:55.501390 sonic INFO pmon#supervisord: xcvrd print("Error: reading sysfs file %s" % **sysfs_sfp_i2c_client_eeprom_path**)
Nov 18 00:57:55.501390 sonic INFO pmon#supervisord: xcvrd NameError: **global name 'sysfs_sfp_i2c_client_eeprom_path' is not defined**

ds952811 committed:
sfputil.py: add extra delays for init sequence

ds952811 committed:
sfputil.py: Monitoring the QSFP-DD initialization state, and reinitiate it if necessary

Signed-off-by: Dante Su <dante.su@broadcom.com>

ds952811 committed:
sfputil.py: drop the QSFPDD software reset from the insertion event

In the current implementation, the SFP change events are actually software
simulated by polling the presence flags, and the INSERTION events will be
be generated upon pmon restart and reboot.

And hence, significant packet drops could be observed upon both pmon restart
and a warm-reboot, as we'll always fire a software reset to the transceivers.

i.e.
It's better to drop the QSFPDD software reset from the insertion event

Signed-off-by: Dante Su <dante.su@broadcom.com>

ds952811 committed:
sfputil.py: clean up

Signed-off-by: Dante Su <dante.su@broadcom.com>

zhenggen-xu commented:
Can you please share the testing log for Innolight and Eoptolink optics 4x100G mode with Silverstone-128x100 HWSKU?  I would like to see the at least link status (up), plug out/in still works, show interface transceiver eeprom -d etc.

zhenggen-xu commented on device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:
```diff
@@ -379,9 +584,34 @@ def get_eeprom_dict(self, port_num):
+                if tbl is not None:
+                    app = 1
+                    hid = int(eeprom_ifraw[1 + offset], 16)
+                    while (hid != 0) and (hid != 0xff):
```
What is this loop doing?  Make a comment here.  Is it possible to have infinity loop?

ds952811 replied:
1. This is to decode the application advertisement and get it displayed on the console.
e.g.
admin@sonic:~$ show interfaces transceiver eeprom
Ethernet0: SFP EEPROM detected
        Application Advertisement:
            1: 400GAUI-8 C2M (Annex 120E) | 400GBASE-DR4 (Cl 124)
            2: 100GAUI-2 C2M (Annex 135G) | 100GBASE-DR (Cl 140)
        Application Selected: 1
        Identifier: QSFP-DD Double Density 8X Pluggable Transceiver
        Interrupt Asserted: Off
        Module State: Ready State
        Revision Compliance: 3.0
        Vendor Name: Eoptolink
        Vendor Part Number: EOLD-134HG-02-M6
        Vendor Revision: 01
        Vendor Serial Number: DKC9A90005

2.  Yes, it could be an infinity loop if the number of supported application mode >=8, will add a sanity check for this.

zhenggen-xu commented on device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:
```diff
@@ -344,11 +393,146 @@ def reset(self, port_num):
+        part = self.inf8628.parse_vendor_pn(buf, 20)['data']['Vendor PN']['value']
+        #log_info("_init_cmis_module: p={0}, id='{1}', name='{2}', part='{3}'".format(port_num, buf[0], name, part))
+        # As of now, init sequence is only necessary for 'InnoLight T-DP4CNT-N00'
+        if name.upper() != 'INNOLIGHT' or part.upper() != 'T-DP4CNT-N00':
```
Can we confirm Eoptolink does not require the init steps?

ds952811 replied:
Yes, it's verified that individual 100G links of Eoptolink could work independently when it's in 1x400G mode, and in fact, we only have 2 out of the 4 links established.
e.g.


admin@sonic:~$ **show interfaces status**
   Interface    Lanes    Speed    MTU     Alias          Vlan    Oper    Admin                                             Type    Asym PFC
'------------  -------  -------  -----  --------  ------------  ------  -------  -----------------------------------------------  ----------'
   Ethernet0    33,34     100G   9100   QSFP1/1  PortChannel4      up       up  QSFP-DD Double Density 8X Pluggable Transceiver         N/A
   Ethernet2    35,36     100G   9100   QSFP1/2  PortChannel4      up       up  QSFP-DD Double Density 8X Pluggable Transceiver         N/A
   Ethernet4    37,38     100G   9100   QSFP1/3         trunk    down       up  QSFP-DD Double Density 8X Pluggable Transceiver         N/A
   Ethernet6    39,40     100G   9100   QSFP1/4         trunk    down       up  QSFP-DD Double Density 8X Pluggable Transceiver         N/A

admin@sonic:~$ **show interfaces transceiver eeprom**
Ethernet0: SFP EEPROM detected
        Application Advertisement:
            **1: 400GAUI-8 C2M (Annex 120E) | 400GBASE-DR4 (Cl 124)**
            2: 100GAUI-2 C2M (Annex 135G) | 100GBASE-DR (Cl 140)
        **Application Selected: 1**
        Identifier: QSFP-DD Double Density 8X Pluggable Transceiver
        Interrupt Asserted: Off
        Module State: Ready State
        Revision Compliance: 3.0
        Vendor Name: **Eoptolink**
        Vendor Part Number: EOLD-134HG-02-M6
        Vendor Revision: 01
        Vendor Serial Number: DKC9A90005

zhenggen-xu commented on device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:
```diff
@@ -344,11 +393,146 @@ def reset(self, port_num):
+        if name.upper() != 'INNOLIGHT' or part.upper() != 'T-DP4CNT-N00':
+            return True
+
+        log_info("PORT {0}: _init_cmis_module".format(port_num))
```
log port_num and module name for better visibility.

ds952811 replied:
Ack

zhenggen-xu commented on device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:
```diff
@@ -344,11 +393,146 @@ def reset(self, port_num):
+        log_info("PORT {0}: _init_cmis_module".format(port_num))
+
+        # Allow 1s for software reset
+        self._write_byte(port_num, self.IDENTITY_EEPROM_ADDR, -1, 26, 0x08)
```
it would be good to wrap below logic into a function to have hwsku, module name as input so easier to expand?  Also, are we deprecating the cmis-init.sh files, if so, remove both of them from the PR.

ds952811 replied:
1. Ack, will do
2. Yes, the use of cmis-init.sh is deprecated, it's a known issue the InnoLight T-DP4CNT-N00 could sometimes failed and we'll need to poll its state for the retry. And the script that we provided is also for concept approval, it's not expected to have it directly integrated as-is.

zhenggen-xu replied:
can you remove both cmis-init.sh from the PR?

ds952811 replied:
Done

zhenggen-xu commented on device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:
```diff
@@ -344,11 +393,146 @@ def reset(self, port_num):
+                if ((int(buf[3], 16) >> 1) & 0x7) != 3:
+                    self.mod_failure[x] += 1
+                # initiate QSFP-DD software reset if failure counter > 2
+                if self.mod_failure[x] > 2:
```
Should we retry inside this loop right away?  Or in the code, when is this retry called?  Another set of unplug/plug events?

ds952811 replied:
1. This is an Eoptolink specific issue, it could be stuck in ModuleLowPwr upon reinsertion, so we'll monitor its state and initiate a software reset when its state != ModuleReady for 2 runs (~3 seconds)
2. It will block the xcvrd for seconds if we retry in this loop right away, this does not seem to be a good idea.

zhenggen-xu replied:
My question was, without later module inserts events, would this code even be called later?  I,E, the module may stuck at bad state for long time?

ds952811 replied:
The get_transceiver_change_event() is invoked by the infinite loop of the xcvrd, and no matter if a insertion/removal event generated or not, it will always poll the module state and trigger a software reset when the failure count > 2.
Note: 
1. The insertion/removal event is generated at line 503, there is an in-memory presence table for the change events
2. Both the module state checker and the QSFPDD config error checker are active when the module is attached, it does not matter if the insertion/removal event is generated or not.  It's outside the if-statement at line 502

zhenggen-xu commented on device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:
```diff
@@ -379,9 +584,34 @@ def get_eeprom_dict(self, port_num):
+                        offset += 4
+                        hid = int(eeprom_ifraw[1 + offset], 16)
+                if len(ret) > 0:
+                    sfp_data['interface']['data']['Application Advertisement'] = ret
```
ret format is a bit strange?

ds952811 replied:
It's an one-line message for the application modes, in the current implementation, it includes host-interface-id and media-interface-id, and it's separated by '|'
e.g.
Ethernet96: SFP EEPROM detected
        Application Advertisement:
            1: 400G CR8 | Copper cable
            2: 200GBASE-CR4 (Clause 136) | Copper cable
            3: 100GBASE-CR4 (Clause 92) | Copper cable
            4: 50GBASE-CR (Clause 126) | Copper cable
            5: IB HDR (Arch.Spec.Vol.2) | Copper cable
            6: 25GBASE-CR CA-L (Clause 110) | Copper cable
            7: IB EDR (Arch.Spec.Vol.2) | Copper cable

zhenggen-xu commented on device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:
```diff
@@ -379,9 +584,34 @@ def get_eeprom_dict(self, port_num):
+                if sfp_data['interface']['data']['Identifier'] not in [type_of_transceiver['18'], type_of_transceiver['19']]:
                     return self.get_qsfp_data(eeprom_ifraw)
 
+                # decode application advertisement
```
Who is using these new attributes?

ds952811 replied:
This is necessary for CMIS3/4 init sequence, for now, we blindly select application 2, but it's not always correct, so I added the support of decoding application advertisement, in the case that we need to add more module-specific init sequence in the future.
Note: The changes in this PR is intentionally revised to InnoLight T-DP4CNT-N00 only, the common CMIS3/4 support is not in our scope.

ds952811 commented:
Hi Xu,

Attached is the UT for your reference, please note while InnoLight is in
application 2, the Eoptolink is in application 1 (default mode), and the
individual 100G links could work independently

Note: I have no idea how to attached the log in the Github, so I respond to
this email, hope it works.

Best Wishes,

Dante (Kuo-Jung) Su


On Thu, Nov 19, 2020 at 6:45 AM zhenggen-xu <notifications@github.com>
wrote:

> Can you please share the testing log for Innolight and Eoptolink optics
> 4x100G mode with Silverstone-128x100 HWSKU? I would like to see the at
> least link status (up), plug out/in still works, show interface transceiver
> eeprom -d etc.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/zhenggen-xu/sonic-buildimage/pull/92#issuecomment-730004881>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMLISKJ5EGMOHIE5NDYZ563SQRFAJANCNFSM4TWTWJFQ>
> .
>

ds952811 committed:
sfputil.py: address review comments

1. Add sanity check to application advertisement decoder to prevent
   infinite loop

2. Create a function wrapper for module-specific CMIS init

Signed-off-by: Dante Su <dante.su@broadcom.com>

ds952811 commented:
The UT result of 'sfputil.py: address review comments':

$ grep pmon /var/log/syslog
......
Nov 19 02:51:05.383744 sonic INFO pmon#xcvrd: Got SFP inserted event
Nov 19 02:51:05.707443 sonic INFO pmon#xcvrd: get sfp info successfully Ethernet246, push to db
Nov 19 02:51:05.707636 sonic INFO pmon#xcvrd: Got SFP inserted event
Nov 19 02:51:06.031486 sonic INFO pmon#xcvrd: get sfp info successfully Ethernet248, push to db
Nov 19 02:51:06.031686 sonic INFO pmon#xcvrd: Got SFP inserted event
Nov 19 02:51:06.355503 sonic INFO pmon#xcvrd: get sfp info successfully Ethernet250, push to db
Nov 19 02:51:06.355850 sonic INFO pmon#xcvrd: Got SFP inserted event
Nov 19 02:51:06.679629 sonic INFO pmon#xcvrd: get sfp info successfully Ethernet252, push to db
Nov 19 02:51:06.679854 sonic INFO pmon#xcvrd: Got SFP inserted event
Nov 19 02:51:07.003463 sonic INFO pmon#xcvrd: get sfp info successfully Ethernet254, push to db
Nov 19 02:51:07.003689 sonic INFO pmon#xcvrd: State transition from 0 to 1
Nov 19 02:51:48.627401 sonic INFO pmon#sfputil.py: PORT 31: INNOLIGHT T-DP4CNT-N00: _init_cmis_module_custom
root@sonic:~#
root@sonic:~#
root@sonic:~#
root@sonic:~#
root@sonic:~# hexdump -C -n 256 -s 0x880 /sys/bus/i2c/devices/40-0050/eeprom
00000880  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000890  00 21 21 25 25 29 29 2d  2d ff 00 00 33 33 33 33  |.!!%%))--...3333|
000008a0  ff ff 22 22 22 22 00 00  00 00 22 22 22 22 00 00  |..""""....""""..|
000008b0  00 00 00 00 10 10 10 10  10 10 10 10 ff 00 00 33  |...............3|
000008c0  33 33 33 ff ff 22 22 22  22 00 00 00 00 22 22 22  |333..""""...."""|
000008d0  22 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |"...............|
000008e0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000900  44 44 44 44 00 00 ff 00  00 00 00 00 0f 00 0f 00  |DDDD............|
00000910  0f 00 0f 0f 00 00 00 00  00 00 3e db 3e 51 3c ca  |..........>.>Q<.|
00000920  3d 09 00 00 00 00 00 00  00 00 9c 40 9c 40 9c 40  |=..........@.@.@|
00000930  9c 40 00 00 00 00 00 00  00 00 3a ad 54 00 31 b3  |.@........:.T.1.|
00000940  35 58 00 00 00 00 00 00  00 00 11 11 11 11 21 21  |5X............!!|
00000950  25 25 29 29 2d 2d ff 00  00 33 33 33 33 ff ff 22  |%%))--...3333.."|
00000960  22 22 22 00 00 00 00 22  22 22 22 00 00 00 00 00  |"""...."""".....|
00000970  11 12 13 14 00 00 00 00  11 12 13 14 00 00 00 00  |................|
00000980
root@sonic:~#

zhenggen-xu commented:
You can drag and drop the txt files into comments or better in PR description,  I would like to see the output of "show interface status",  "show interface transciever eeprom -d" with multiple optical modules in place.

ds952811 committed:
device/celestica/x86_64-cel_silverstone-r0: drop cmis-init.sh

Signed-off-by: Dante Su <dante.su@broadcom.com>

ds952811 commented:
1. Can you remove both cmis-init.sh from the PR?
Done, please refer to commit bf811ae9f3d3fd2869e5ba3adfa1b7afbf517924
2. Attached is the UT result
[QSFPDD_UT.log](https://github.com/zhenggen-xu/sonic-buildimage/files/5564727/QSFPDD_UT.log)

Note:

- Eoptolink: Only 2 out of 4 links were UP, as we only had 2 cables physically connected, and it's also evidence that all of the 100G links on the Eoptolink could work independently in the default application mode (i.e. AppSel=1, 1x400G mode)
- InnoLight: Please check the last result of 'show interface status', while the Ethernet242 was down, Ethernet240, Ethernet244 and Ethernet246 were still up. And it was running at application 2, which is 4x100G mode.

ds952811 commented:
Not sure why the inline response is not displayed on the conversation page, here is my response to the last question:

> My question was, without later module inserts events, would this code even be called later? I,E, the module may stuck at bad state for long time?


The get_transceiver_change_event() is invoked by the infinite loop of the xcvrd, and no matter if a insertion/removal event generated or not, it will always poll the module state and trigger a software reset when the failure count > 2.
Note:

The insertion/removal event is generated at line 503, there is an in-memory presence table for the change events
Both the module state checker and the QSFPDD config error checker are active when the module is attached, it does not matter if the insertion/removal event is generated or not. It's outside the if-statement at line 502
